python/ray/data/_internal/logical/rules/limit_pushdown.py

```diff
@@ -0,0 +1,128 @@
+import copy
+from collections import deque
+from typing import Iterable, List
+
+from ray.data._internal.logical.interfaces import LogicalOperator, LogicalPlan, Rule
+from ray.data._internal.logical.operators.one_to_one_operator import (
+    AbstractOneToOne,
+    Limit,
+)
+from ray.data._internal.logical.operators.read_operator import Read
+
+
+class LimitPushdownRule(Rule):
+    """Rule for pushing down the limit operator.
+
+    When a limit operator is present, we apply the limit on the
+    most upstream operator that supports it. Notably, we move the
+    Limit operator downstream from Read op, any other non-OneToOne operator,
+    or any operator which could potentially change the number of output rows.
+
+    In addition, we also fuse consecutive Limit operators into a single
+    Limit operator, i.e. `Limit[n] -> Limit[m]` becomes `Limit[min(n, m)]`.
+    """
+
+    def apply(self, plan: LogicalPlan) -> LogicalPlan:
+        optimized_dag = self._apply_limit_pushdown(plan.dag)
+        optimized_dag = self._apply_limit_fusion(optimized_dag)
+        return LogicalPlan(dag=optimized_dag)
+
+    def _apply_limit_pushdown(self, op: LogicalOperator) -> LogicalOperator:
+        """Given a DAG of LogicalOperators, traverse the DAG and push down
+        Limit operators, i.e. move Limit operators as far upstream as possible.
+
+        Returns a new LogicalOperator with the Limit operators pushed down."""
+        # Post-order traversal.
+        nodes: Iterable[LogicalOperator] = deque()
+        for node in op.post_order_iter():
+            nodes.appendleft(node)
+
+        while len(nodes) > 0:
+            current_op = nodes.pop()
+
+            # If we encounter a Limit op, move it upstream until it reaches:
+            # - Read operator
+            # - A non-AbstractOneToOne operator (e.g. AbstractAllToAll)
```

Comment:
Read operator is not one-to-one either?

Reply:
in our current abstraction, we have Read extend AbstractMap with no input dependency, which extends AbstractOneToOne. so we have to separate this as a case

Reply:
I see. But in theory, I don't think Read should extend AbstractMap. We can change this in the next PR.

Reply:
agreed, i have also left a comment to clarify why we separated this into its own case.